src/main/java/com/shrralis/ssdemo1/security/AuthProvider.java

```diff
@@ -70,18 +82,21 @@ private UserDetails retrieveUserDetails(String login) {
      * @throws BadCredentialsException if invalid password was entered
      */
     private void additionalAuthenticationChecks(
-		    UserDetails userDetails,
+		    AuthorizedUser userDetails,
 		    UsernamePasswordAuthenticationToken authentication) {
         if (authentication.getCredentials() != null) {
             String password = authentication.getCredentials().toString();
 
 	        if (password.length() < User.MIN_PASSWORD_LENGTH
 			        || !passwordEncoder.matches(password, userDetails.getPassword())) {
-                throw new BadCredentialsException("Bad credentials");
+		        userDetailsService.increaseUserFailedAttempts(userDetails);
+		        throw new CitizenBadCredentialsException(
+				        userDetails.getUsername(), userDetails.getFailedAuthCount() + 1);
             }
         } else {
-            throw new BadCredentialsException("Bad credentials");
+	        throw new BadCredentialsException(userDetails.getUsername());
```

Comment:
Inverse this if.
if ( == null) { BadCredentialsException}
...

Reply:
O.K., I missed that, thanks